### ISSUE TITLE
Make backtick strings usable for SQL parameters

### ISSUE THREAD

**Issue by jclark** [ballerina-platform/ballerina-spec#442]
Basic idea is that:

```
var x = `x${expr1}y${expr2}z`
```
desugars to

```
var x = {
    strings: ["x", "y", "z" ],
    insertions: [expr1, expr2]
};
```
except that the `strings` array is immutable.

There would be a built-in abstract object type `ParameterizedString<T>` where `T` is the type of the `insertions` field and must be a list type. For the SQL case, T would be `sql:Value[]`.

This is similar to how ECMAScript tagged template literals work.

**Comment by sanjiva** [comment]
In general though the insertions array would be heterogeneous right? T would end up with a union type in many cases.

I guess that's ok - for SQL functions the type would be `ParameterizedString<sql:Value[]>`.

The `[]` bit is rather redundant .. should it not be just `T`? The `insertions` field would (eventually) be of type `readonly T[]`.

**Comment by jclark** [comment]
The type inferred for T would be tuple type, e.g. for `x${1}y${true}z` it would be `[int, boolean]`. When you analyze the string, different insertion points may allow different types to be inserted. For example in the XML case e.g. `<doc att="${E1}">${E2}</doc>` E2 should allow arbitrary xml values but E1 shouldn't.

**Comment by sanjiva** [comment]
Duh yes I had misunderstood the previous comment from you. So in general T would be a tuple type with the array case being the rare case (like SQL).

**Comment by jclark** [comment]
Here's a scheme that will actually work.  Somewhere in langlib, we define:

```
branded type BacktickString record {|
   readonly string[] strings;
|};
```
Then

```
 `x${expr1}y${expr2}z`
```

desugars to

```
[BacktickString(["x", "y", "z" ]), expr1, expr2]
```

The first argument to the query method of the SQL client will then have type:

```
[BacktickString, sql:Value...]
```

**Comment by jclark** [comment]
We should build this on top of our current [design](https://github.com/ballerina-platform/ballerina-spec/blob/master/lang/proposals/distinct/distinct.md) for nominal typing.

First, we have a type defined in lang.object. 

```
// module lang.object
public type RawTemplate branded object {
   public readonly string[] strings;
   public (any|error) insertions[];
};
```

The sql module would define a subtype of this:

```
// module sql
public type Value ()|int|float|decimal|string|xml|TypedValue;
public type TypedValue distinct object {
   public readonly value;
};
public type SmallIntValue distinct object {
   *TypedValue;
   public readonly int value;
   public function init(int value) { this.value = value; }
};

public abstract type QueryString object {
  *object:RawTemplate;
  public readonly Value insertions[];
};

public type Client abstract client object {
  public remote function query(QueryString|string q, typedesc<record {}> rowType = <>)
      returns stream<rowType, Error> = external;
};
```

Now we can do

```
jdbc:Client client;
stream<Customer,*> stm = client->query(`x${expr1}y${expr2}z`);
```

and it should desugar to:

```
jdbc:Client client;
type _O object {
   public readonly string[] strings = ["x", "y", "z"];
   public readonly sql:Value insertions[];
   public function __init(sql:Value val1, sql:Value val2) {
     self.insertions = [val1, val2];
  }
};
stream<Customer,*> stm = client->query(new _O(expr1, expr2));
```

**Comment by pubudu91** [comment]
A couple of questions regarding this. Why does the type of `strings` have to be `readonly & string[]`? 

Also, shouldn't we take the signature of the user defined init method in to consideration when type checking? The concern is, if there are arbitrary parameters defined in the init method, we won't be able to desugar the type-init-expr.

**Comment by jclark** [comment]
The array of strings should be constructed once at compile-time.

I don't understand the second question: the signature of the `init` method is not part of the type of an object.

**Comment by pubudu91** [comment]
> The array of strings should be constructed once at compile-time. Apart from being the right thing to do, it is useful for `batchExecute`.

If it's to ensure the string array is constructed only once, shouldn't the `strings` field be defined as 
```ballerina
public readonly string[] strings;
```
instead of 
```ballerina
public (readonly & string[]) strings;
```
?
> I don't understand the second question: the signature of the `init` method is not part of the type of an object.

I meant something like the following:
```ballerina
public type Template object {
    *'object:RawTemplate;
    public anydata[] insertions = [];
    int foo = 0;

    function __init(int foo) {
        self.foo = foo;
    }
};

public function main() {
    string name = "John Doe";
    Template t = `Hello ${name}!`;
}
```
Basically the user defines a subtype of `RawTemplate` and defines an init method with an arbitrary signature.

**Comment by jclark** [comment]
`readonly & string[]` is the correct type for an immutable array of strings.

The `init` method is not part of an object’s type. See last para of https://ballerina.io/spec/2020R1/#section_5.4.2.5

**Comment by pubudu91** [comment]
> `readonly & string[]` is the correct type for an immutable array of strings.

Sorry, should've phrased this in a better way. Yes, the value should be immutable, agreed. My question is whether the `strings` field of the `RawTemplate` object itself should be immutable as well after the value is constructed? i.e., after assigning a raw template literal to `rt`, can I do `rt.strings = aDifferentReadonlyArray;` later on?

> The `init` method is not part of an object’s type. See last para of https://ballerina.io/spec/2020R1/#section_5.4.2.5

No, I meant, when desugaring the literal to an object value, we run in to a problem with missing arguments to the type-init-expr if we allow the user to define init methods with arbitrary parameters. For example, currently I'm desugaring a literal as follows:
```ballerina
'object:RawTemplate rt = `Hello ${name}`;

// the literal part is desugared to
{
    'object:RawTemplate $template$ = new;
    $template$.strings = ["Hello "];
    $template$.insertions = [name];
}

// And the (simplified version) desugared statement would look like the following:
'object:RawTemplate rt = $template$;
```
Now support a user defines a subtype of `RawTemplate`, with an explicit init method.
```ballerina
public type CustomTemplate object {
    *'object:RawTemplate;

    function __init(int foo, string bar) {
        // some init code
    }
};

// then, somewhere in the code, a template literal is assigned to a var of this type.
CustomTemplate ct = `Hello ${name}`;

// Now attempting to desugar the above, 
{
    CustomTemplate $template$ = new(?, ?); // A problem here since this type-init-expr expects arguments
    $template$.strings = ["Hello "];
    $template$.insertions = [name];
}
```
This is why I was wondering whether we should disallow assigning raw template literals to subtypes of `RawTemplate` with user defined init methods.

**Comment by jclark** [comment]
Let me answer your second question first.

The spec's current terminology for object types makes things a bit confusing. So in this reply I will adopt the terminology from https://github.com/ballerina-platform/ballerina-spec/issues/535#issuecomment-639237263

The goal of the desugaring is:
- allow control over the types of the inserted expressions
- allow a value that comes from desugaring a RawTemplate to be distinguished from other values

The goal is *not* to allow the user to supply a user-defined *class* for the raw template.

So what you need to infer is an object *type* (which is a subtype of object:RawTemplate,) not an object *class*. The context provides you with a type, which gives you the precise type for the `insertions` field. The desugaring then defines its own class which implements that type.

Suppose we have
```
type MyParamString abstract object {
    (readonly & string[]) strings;
    MyParamType[] params;
};

function foo(NameType name) returns MyParamString {
   return `Hello ${name}!`;
}
```

Then `foo` would desugar to something like:

```
// at module level
const $strings$ = ["Hello ", "!"];
// a subtype of MyParamString
type $Template$ = object {
   *object:RawTemplate;
   (readonly & string[]) strings = $strings$;
   [MyParamType] insertions;
   __init(MyParamType arg) { this.insertions = [arg]; }
};

function foo(NameType name) returns MyParamString {
   return new $Template$(name);
}
```

This means that contextually required type mustn't have any methods or required fields other than `strings` and `insertions`.

On your 1st question, there is no need to require the strings field to be immutable, since the field isn't shared between evaluations of the raw template.  The user can make it immutable if they want by making their raw template subtype have the field readonly (since a field `readonly string[] strings;` is a subtype of `(readonly & string[]) strings;`_. This approach allows the insertions field also to be readonly if the type of the insertions are all readonly.

**Comment by pubudu91** [comment]
I think I got it now. Reiterating to see if my understanding is correct.
- The template literals are meant to work just with object types which are subtypes of `object:RawTemplate`. Prevent the user from using it in class contexts.
- Furthermore, the only members that subtype should have are the two fields: `strings` and `insertions`

Going by the above reasoning, shouldn't the langlib definition for `RawTemplate` given in https://github.com/ballerina-platform/ballerina-spec/issues/442#issuecomment-619482206 also be a type? 

> On your 1st question, there is no need to require the strings field to be immutable, since the field isn't shared between evaluations of the raw template. The user can make it immutable if they want by making their raw template subtype have the field readonly (since a field readonly string[] strings; is a subtype of (readonly & string[]) strings;_. This approach allows the insertions field also to be readonly if the type of the insertions are all readonly.

Thank you, this clarifies the questions I had regarding the fields.

**Comment by jclark** [comment]
Yes, `object:RawTemplate` should be abstract.

The user isn't directly prevented from using it in a class context: it's just that it's only the type that is inferred. If the class is such that you cannot define a class with the right type, then it will be an error.

**Comment by pubudu91** [comment]
Thanks! That clears up the doubts. Will modify the behaviour to conform to the behaviour described.

**Comment by niveathika** [comment]
@pubudu91,
For 
```
var x = `x${expr1}y${expr2}z`
```
when desugaring x and z are mandatory, isn't it? In the sense, if users give following raw template, without x and z. 
```
var x = `${expr1}y${expr2}`
var x = [
strings = ["", y, ""]
insertions = [expr1, expr2]
]
```
When desugaring we should have empty strings in their place, else it will be ambiguous from which point we have to start to combine the query.

**Comment by jclark** [comment]
Yes, you need to put the empty strings in. If there are n expressions, there should be n+1 strings.
